src/blockchain/lending/market.ts

```diff
@@ -0,0 +1,237 @@
+import { BaseProvider } from '@ethersproject/providers'
+import { parseEther } from '@ethersproject/units'
+import { Abi, ADDRESSES, LENDING_LIQUIDITY_POOL } from '@xtoken/abis'
+import { BigNumber, Contract } from 'ethers'
+
+import { getSignerAddress } from '../utils'
+
+import { getMarkets, Markets } from './helper'
+
+// --- Market functions ---
+
+/**
+ * Add xAsset collateral to a Lending Market
+ * @param marketName name of the market
+ * @param amount amount without decimals
+ * @param provider
+ * @returns
+ */
+export const collateralize = async (
+  marketName: Markets,
+  amount: string,
+  provider: BaseProvider
+) => {
+  const markets = await getMarkets(provider)
+  const market = markets[marketName]
+  const inputAmount = parseEther(amount)
+  const address = await getSignerAddress(provider)
+  const approvedAmount = await getApprovedAmountXAsset(
+    marketName,
+    address,
+    provider
+  )
+  if (approvedAmount.lt(inputAmount)) {
+    return Promise.reject(
+      new Error('Please approve the tokens before adding collateral')
+    )
+  }
+  return market.collateralize(inputAmount)
+}
+
+/**
+ * Withdraw xAsset collateral from a Lending Market
+ * @param marketName name of the market
+ * @param amount amount without decimals
+ * @param provider
+ * @returns
+ */
+export const withdrawCollateral = async (
+  marketName: Markets,
+  amount: string,
+  provider: BaseProvider
+) => {
+  const markets = await getMarkets(provider)
+  const market = markets[marketName]
+  const inputAmount = parseEther(amount)
+  return market.withdraw(inputAmount)
+}
+
+/**
+ * Get xAsset collateral deposited in a Lending Market for an address
+ * @param marketName name of the market
+ * @param provider
+ * @param address optional address - checks current signer if not provided
+ * @returns
+ */
+export const getCollateral = async (
+  marketName: Markets,
+  provider: BaseProvider,
+  address?: string
```

Comment:
What's the reason for this to be optional? Can we call `market.collateral()` if `address` is `undefined`?

Reply:
If address is passed to the function, we get that address's collateral. 
If it's not, we get the provider's signer address's collateral.